src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRunSymbolNodeFactory.cs

```diff
@@ -45,99 +41,228 @@ public sealed class ReadyToRunSymbolNodeFactory
+            {
+                ModuleToken = moduleToken;
+                SignatureContext = signatureContext;
+            }
+
+            public bool Equals(ModuleTokenAndSignatureContext other)
+            {
+                return ModuleToken.Equals(other.ModuleToken)
+                    && SignatureContext.Equals(other.SignatureContext);
+            }
+
+            public override bool Equals(object obj)
+            {
+                return obj is ModuleTokenAndSignatureContext other && Equals(other);
+            }
+
+            public override int GetHashCode()
             {
-                stringNode = new StringImport(_codegenNodeFactory.StringImports, token, signatureContext);
-                _importStrings.Add(token, stringNode);
+                return ModuleToken.GetHashCode() ^ (SignatureContext.GetHashCode() * 31);
             }
-            return stringNode;
         }
 
-        private readonly Dictionary<ReadyToRunHelperId, Dictionary<object, ISymbolNode>> _r2rHelpers = new Dictionary<ReadyToRunHelperId, Dictionary<object, ISymbolNode>>();
+        private NodeCache<ModuleTokenAndSignatureContext, ISymbolNode> _importStrings;
 
-        public ISymbolNode ReadyToRunHelper(ReadyToRunHelperId id, object target, SignatureContext signatureContext)
+        public ISymbolNode StringLiteral(ModuleToken moduleToken, SignatureContext signatureContext)
+        {
+            return _importStrings.GetOrAdd(new ModuleTokenAndSignatureContext(moduleToken, signatureContext));
+        }
+
+        private struct ReadyToRunHelperKey
         {
-            if (!_r2rHelpers.TryGetValue(id, out Dictionary<object, ISymbolNode> helperNodeMap))
+            public readonly ReadyToRunHelperId Id;
+            public readonly object Target;
+            public readonly SignatureContext SignatureContext;
+
+            public ReadyToRunHelperKey(ReadyToRunHelperId id, object target, SignatureContext signatureContext)
+            {
+                Id = id;
+                Target = target;
+                SignatureContext = signatureContext;
+            }
+
+            public bool Equals(ReadyToRunHelperKey other)
+            {
+                return Id == other.Id
+                    && Target.Equals(other.Target)
+                    && SignatureContext.Equals(other.SignatureContext);
+            }
+
+            public override bool Equals(object obj)
             {
-                helperNodeMap = new Dictionary<object, ISymbolNode>();
-                _r2rHelpers.Add(id, helperNodeMap);
+                return obj is ReadyToRunHelperKey other && Equals(other);
             }
 
-            if (helperNodeMap.TryGetValue(target, out ISymbolNode helperNode))
+            public override int GetHashCode()
             {
-                return helperNode;
+                return Id.GetHashCode() ^ (SignatureContext.GetHashCode() * 31);
             }
```

Comment:
`Target` is missing in the hashcode computation

Reply:
Added.